torch/csrc/autograd/python_engine.cpp

```diff
@@ -3,13 +3,33 @@
 #include "torch/csrc/autograd/engine.h"
 #include "torch/csrc/THP.h"
 #include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/utils/auto_gil.h"
 
 using namespace torch::autograd;
 
 struct THPEngine {
     PyObject_HEAD
 };
 
+struct PythonEngine : public Engine {
+  virtual void thread_main(ReadyQueue& queue) override {
+    AutoGIL gil;
+    AutoNoGIL no_gil;
```

Comment:
Can you please add a comment with all that? It looks a weird when you don't have the context.